### ISSUE TITLE
Feature Request: Postman Environment Permissions

### ISSUE THREAD

**Issue by mhertogs** [a85/POSTMan-Chrome-Extension#4290]
## App Details:

```
Postman for Mac
Version 6.0.9-canary02
OS X 16.7.0 / x64
```

## Feature Request:
I'm excited about the ability to share dynamic environments in Postman 6.x but with this new feature, it also makes it more difficult to maintain immutable environments for all users. As the administrator of my company's Postman, it is useful to be able to create collections and environments that are not editable by other users - this allows the collections and environments to serve as immutable public documentation for our APIs.

In summary, it would be useful to be able to add permissions for environments - i.e. you should be able to designate environments as "Can View" or "Can Edit" similar to collections.

**Comment by harryi3t** [comment]
@mhertogs Thanks for reaching out to us. I have marked this as a feature request. Will keep you posted on the status.

**Comment by Telaran** [comment]
If possible I would request a bit more granularity on the environments.
As of now we cannot use the Workspace (and I couldn't wait for this feature since you mentioned it)

## Usage in our company
Currently I'm the maintainer of most of our collections and also environments.
I'm using the Environments with "Pre-Request Script" and also "Tests" almost everywhere.
- Login -> Set Access Token in Environment
- POST Requests -> Set ID of new Object in Environment
- Prefill some environments with generic values (for Testing)
- etc

This means that many Environment-Variables are necessary for the team-members, but shouldn't be synced and stay within the team-member system.

## Request
I would like to have a "_can synced_" flag. So this means the user can keep his variables, but they will not be overwritten or synced by Workspace.

### PS
Except this Sync Issue I really love the new feature. Finally it's possible to add new values and the team-members will get them (so they don't have to add manually new introduced variables).
Also the Workspace is a long waited feature (we have a big amount of collections and environments). Keep up the good work and hopefully I can soon switch to Workspace with the team.

**Comment by mhertogs** [comment]
Agreed with @Telaran. My initial request for environment permissions is bare minimum but the ideal would be to have environments be able to have two tiers of variables - shared variables that are kept in sync and variables unique to this individual using the environment.

**Comment by harryi3t** [comment]
@Telaran Assuming environment levels permissions are added, what use-cases are still not covered which requires `non-syncable` environments?

Users can always mark their env as read-only then no one will be able to mutate them.

If you don't want others to see the values, then maybe you could share a read-only env with only keys and no values (kind of a template). Individual users can clone the env and move it into their personal workspace and then use it there (Also, sharing the collection to the same personal workspace).

If you turn off sync on just one env, then if you sign out (or log into new machine) that data will be lost.

**Comment by mhertogs** [comment]
## Example Use Case
Imagine I have a webservice deployed into three separate deployment clusters (dev/staging/prod) and each deployment has a different host and port for the webservice. These host/port combos in each cluster are subject to change and I want to be able to update the environment.

I create the following Postman environments. 
- Dev Env: host=1.2.3.4, port:34567
- Staging Env: host=2.3.4.5, port: 45678
- Prod Env: host=3.4.5.6, port: 56789

Note: I would like these variables to be immutable and synced. Immutable so that no one user accidentally fat-fingers the port and ruins the environment for all users, and synced so if the host or port changes in the staging cluster for the webservice, I can update the environment for everyone using it. 

Now examine the two solutions proposed:

**1. The environments are set to read-only**
Individual users writing collections against this webservice now lose the ability to write to that environment. Therefore, they can't use environment to store their personal access token, use pm.environment.set to save value of a response for use in the next request, etc. This makes it difficult to use all of the nice features that Postman provides.

If copying it to their individual workspace allows them to write new variables to the environment, then we are essentially back to environment templates in 5.x because they would not sync at that point.

**2. The environments allow synced variables and non-synced variables**
The host and port variables are read-only and synced but the rest of the environment variables that an individual user creates can either be local or shared in that environment. Therefore, the user can easily use ephemeral environment variables during their session.

## Summary
However, @harryi3t raises a good point that Postman would still need to store your individual environment variables or else you would lose them. I think maybe shared environments in a workspace should have workspace variables with permissions that are visible to everyone AND individual variables per user subscribed to the workspace.

But I think that adding permissions is a good first step either way.

**Comment by Telaran** [comment]
@mhertogs thanks for your efforts. I'll can confirm that this explains my use case.

### My Workspace, Team Workspace, the "Share" workaround
Maybe there is just a "different understanding" what Workspaces are for or not (so maybe we want to use it as never intended by Postman?)
In our team only two (should) maintain the collection. All other team-member consume the provided collection, etc.
Therefore it's preferred, that they don't have to invest a lot of time in handling Postman (Create for all 30+ Projects own "My Workspaces" with Shared Collections/Environments). It's more or less my job to give them proper Workspaces (or as currently -> Collections and Environments).

Still they do need "writable" Environment, because of our regulations.
They create their application (Web, App, etc) and if anything crash or don't work as expected they have to test their use case in Postman. We (the Backend Developer) will not look on a case which can't be reproduced in Postman. 
This requires that Workspace shouldn't be read-only and it's better if we have Team Workspaces (so there will be no differences and new introduced variables in the Environment are already there)

#### Example on the current state with Workspaces (which isn't ideal)
- Environment Variable: _AccessToken = DEV2ABC_
- Developer 1 -> Call the Login Request with "Dev1 Account" -> Tests-Part running and set Environment Variables.
- Environment Variable: _AccessToken = DEV1ABC_
- Environment Variable will be synced to all team member
- Developer 2 -> Call an [User] Request which need AccessToken -> Because Environment Variable is synced, the User is no longer the "Dev2 Account" but "Dev1 Account" -> Request failed or result are different than Developer 2 expected -> Developer 2 will most likely file an issue to us or waste time to find the issue (or worse, login again and than Developer 1 will have the issues)

The "workaround" as now would be: Create for each Team-Member his own Environment (for all 2-3 Stages) and then we're back at 10+ Environments which each developer have to take care of (not use the wrong otherwise it will be even more distracting as before) or they have do create their own "My Workspaces" (and waste their precious time on maintain something which could be already done by me)

But as mentioned. Maybe we just want to use the Workspaces differently as Postman intended and we should stay on "the good old chaotic way".

**Comment by ghost** [comment]
Totally agree with this feature request.

Also have similar problem as @Telaran described. Support for private environments for each developer is a must for us. Otherwise with 10 developers each with 2-3 private environments becomes too cluttered.

**Comment by mhertogs** [comment]
Hey any updates on this issue? It is blocking my entire company from moving to Postman 6.0

**Comment by bmesin** [comment]
Please implement this feature asap because it is giving a lot of users headaches.

1. env should be editable ONLY by users who can edit original workspace (env point of origin)
2. original workspace should be visible somewhere (in manage environment window)
3. if I share env to another team workspace and then copy it, then this new team workspace becomes origin for copied env (shared env cannot be edited in this new workspace)
4. every key (entry) in an env we can be marked as "do not sync value" - which means users can modify its value from anywhere and it will not get synced (even from original workspace)
5. keys marked as "do not sync" could not be deleted - only their value would not be synced

I think that this would resolve lots of problems.

I am also postman admin in my company and because of missing "read only" feature for environments I have to give extra tutorials to team members and spend time monitoring who did what.

**Comment by rebrovoleksii** [comment]
@harryi3t Hi there, the company I'm working on wanted to purchase Pro version of Postman but we didn't go for this only because of environment permissions were missing according to our use cases.
Could you please advise what is the status of the current feature request?

**Comment by harryi3t** [comment]
@rebrovoleksii This feature is on our radar and we will be solving this by using something similar to role-based access control ([RBAC](https://en.wikipedia.org/wiki/Role-based_access_control)). I can't commit to the timelines but as soon as this is on our immediate roadmap, I will update the thread.
Thanks for using Postman!

**Comment by rebrovoleksii** [comment]
@harryi3t thanks for update!

**Comment by snowe2010** [comment]
I don't understand why you would be solving this with RBAC. this solution really just needs a third way to set variables. `postman.setPersonalVariable()`. You can get to the rest of the issues later. This will solve the majority of peoples' issues immediately. Yes pretty much everyone wants immutable Environments, because that makes sense, but this is a simpler solution that can be implemented immediately instead of waiting on some long term RBAC solution.

**Comment by bmesin** [comment]
@harryi3t RBAC is wrong way of handling this. It will add extra complexity where it is not needed and prolong fixing this with no real benefit. You guys will have to write new tutorials, solve a bunch of new issues on roles management and permissions ... not worth it in my opinion.
This should be a clean fix. KISS.

**Comment by a85** [comment]
There are two separate issues discussed here: one is about working concurrently on personal values  separately in a team context and then the ability is to edit the environment itself. We are introducing the notion of sessions to help with the former. More details here: https://community.getpostman.com/t/sessions-in-postman/1925

**Comment by a85** [comment]
RBAC Is going to be a more generic system applicable to not just environments but other resources like collections, workspaces etc.

**Comment by a85** [comment]
Postman supports sessions now so each developer can have their own local copy of values. Through that developers can work on local copies of the values they use while shared copies can be stored in initial values of variable definitions. Local values (seen as current values) don't sync to Postman servers or in Workspaces: 

<img width="737" alt="Screenshot 2019-05-04 18 58 27" src="https://user-images.githubusercontent.com/653409/57187077-a09c5980-6e9e-11e9-95af-e2e48b31100e.png">

**Comment by bmesin** [comment]
Local sessions are fine and current values also, but currently every developer can change initial value and sync it back to Postman server and in workspace.
Devs do that by accident most of the times and currently I cannot forbid them to do that.

Is there a way to lock initial values so only collection owner can change initial values and every one else only current values?

**Comment by snowe2010** [comment]
yeah, sessions don't solve the issue at all. It's still present and causing problems constantly.

**Comment by akshay-patel-dunelm** [comment]
@a85 Thanks for the update, however, sessions really haven't solved anything for us.

Sessions aren't sufficiently explained in the application. And due to this users end up changing BOTH values when something isn't working right. This can break the workspace for multiple users.

We have information in our environment variables which is vital to our API's and must be kept up to date. But due to the way postman works, I have to provide instructions on how to navigate to each environment and click 'Reset All' for each member of the team. And since the Persist All button is a few pixels away I inevitably have to end up fixing variables multiple times due mis-clicks. Again potentially breaking our workspace for many users until I'm able to fix it.

From my perspective, and in my opinion, there are two main problems:
- Any user in a shared workspace can update the Initial Value for all other users, potentially breaking things.
- Resetting all environment variables is not straightforward for users who don't regularly use the Manage Environments functionality, or understand Sessions.

More importantly a user should not be able to break an entire workspace for other people without realising.

**Comment by sankalp0o** [comment]
Thanks for the feedback everyone! Setting up roles and permissions to environment is on our radar and we're tracking it at #6596.
